common/src/intrusive_circular_list.h

```diff
@@ -0,0 +1,662 @@
+/*
+ Copyright (C) 2019 Kristian Duske
+
+ This file is part of TrenchBroom.
+
+ TrenchBroom is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ TrenchBroom is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with TrenchBroom. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef TRENCHBROOM_INTRUSIVE_CIRCULAR_LIST_H
+#define TRENCHBROOM_INTRUSIVE_CIRCULAR_LIST_H
+
+#include <cstddef>
+#include <utility>
+
+/**
+ * This structure contains the link information stored in list items.
+ *
+ * @tparam T the type of the list items
+ */
+template <typename T>
+class intrusive_circular_link {
+    static_assert(!std::is_pointer<T>::value, "intrusive lists do not accept pointer arguments");
+    template <typename, typename> friend class intrusive_circular_list;
+private:
+    T* m_next;
+    T* m_previous;
+public:
+    /**
+     * Creates a new link for the given list item. Sets both next and previous to the given item so that the
+     * link forms a self loop.
+     *
+     * @param item the list item for this link
+     */
+    explicit intrusive_circular_link(T* item) :
+    m_next(item),
+    m_previous(item) {}
+
+    /**
+     * Returns this link's predecessor.
+     */
+    T* previous() const {
+        return m_previous;
+    }
+
+    /**
+     * Returns this link's successor.
+     */
+    T* next() const {
+        return m_next;
+    }
+private:
+    /**
+     * Sets this link's predecessor.
+     */
+    void set_previous(T* previous) {
+        m_previous = previous;
+    }
+
+    /**
+     * Sets this link's successor.
+     */
+    void set_next(T* next) {
+        m_next = next;
+    }
+
+    /**
+     * Flips this link by exchanging its predecessor and its successor.
+     */
+    void flip() {
+        using std::swap;
+        swap(m_next, m_previous);
+    }
+};
+
+/**
+ * A circular list that stores its links inside of the list items.
```

Comment:
Also needs to mention that its destructor deletes the items in the list.

Reply:
Yes, good point.